starky2/src/get_challenges.rs

```diff
@@ -0,0 +1,367 @@
+            config,
+            degree_bits,
+        )
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+pub(crate) fn get_challenges_target<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    S: Stark<F, D>,
+    const D: usize,
+>(
+    builder: &mut CircuitBuilder<F, D>,
+    stark: &S,
+    trace_cap: &MerkleCapTarget,
+    permutation_zs_cap: Option<&MerkleCapTarget>,
+    quotient_polys_cap: &MerkleCapTarget,
+    openings: &StarkOpeningSetTarget<D>,
+    commit_phase_merkle_caps: &[MerkleCapTarget],
+    final_poly: &PolynomialCoeffsExtTarget<D>,
+    pow_witness: Target,
+    config: &StarkConfig,
+) -> StarkProofChallengesTarget<D>
+where
+    C::Hasher: AlgebraicHasher<F>,
+{
+    let num_challenges = config.num_challenges;
+
+    let mut challenger = RecursiveChallenger::<F, C::Hasher, D>::new(builder);
+
+    challenger.observe_cap(trace_cap);
+
+    let permutation_challenge_sets = permutation_zs_cap.map(|permutation_zs_cap| {
+        let tmp = get_n_permutation_challenge_sets_target(
+            builder,
+            &mut challenger,
+            num_challenges,
+            stark.permutation_batch_size(),
+        );
+        challenger.observe_cap(permutation_zs_cap);
+        tmp
+    });
+
+    let stark_alphas = challenger.get_n_challenges(builder, num_challenges);
+
+    challenger.observe_cap(quotient_polys_cap);
+    let stark_zeta = challenger.get_extension_challenge(builder);
+
+    challenger.observe_openings(&openings.to_fri_openings());
+
+    StarkProofChallengesTarget {
+        permutation_challenge_sets,
+        stark_alphas,
+        stark_zeta,
+        fri_challenges: challenger.fri_challenges::<C>(
+            builder,
+            commit_phase_merkle_caps,
+            final_poly,
+            pow_witness,
+            &config.fri_config,
+        ),
+    }
+}
+
+impl<const D: usize> StarkProofWithPublicInputsTarget<D> {
+    pub(crate) fn get_challenges<
+        F: RichField + Extendable<D>,
+        C: GenericConfig<D, F = F>,
+        S: Stark<F, D>,
+    >(
+        &self,
+        builder: &mut CircuitBuilder<F, D>,
+        stark: &S,
+        config: &StarkConfig,
+    ) -> StarkProofChallengesTarget<D>
+    where
+        C::Hasher: AlgebraicHasher<F>,
+    {
+        let StarkProofTarget {
+            trace_cap,
+            permutation_zs_cap,
+            quotient_polys_cap,
+            openings,
+            opening_proof:
+                FriProofTarget {
+                    commit_phase_merkle_caps,
+                    final_poly,
+                    pow_witness,
+                    ..
+                },
+        } = &self.proof;
+
+        get_challenges_target::<F, C, S, D>(
+            builder,
+            stark,
+            trace_cap,
+            permutation_zs_cap.as_ref(),
+            quotient_polys_cap,
+            openings,
+            commit_phase_merkle_caps,
+            final_poly,
+            *pow_witness,
+            config,
+        )
```

Comment:
I think it may be a bit nicer to inline `get_challenges_target`, to avoid the extra function with the long param list. If we skip the destructuring (can keep it if you prefer though), the combined function is only slightly longer than `get_challenges_target`, overall it saves 48 lines -

```suggestion
        let proof = &self.proof;
        let opening_proof = &proof.opening_proof;
        let num_challenges = config.num_challenges;

        let mut challenger = RecursiveChallenger::<F, C::Hasher, D>::new(builder);

        challenger.observe_cap(&proof.trace_cap);

        let permutation_challenge_sets =
            proof.permutation_zs_cap.as_ref().map(|permutation_zs_cap| {
                let tmp = get_n_permutation_challenge_sets_target(
                    builder,
                    &mut challenger,
                    num_challenges,
                    stark.permutation_batch_size(),
                );
                challenger.observe_cap(permutation_zs_cap);
                tmp
            });

        let stark_alphas = challenger.get_n_challenges(builder, num_challenges);

        challenger.observe_cap(&proof.quotient_polys_cap);
        let stark_zeta = challenger.get_extension_challenge(builder);

        challenger.observe_openings(&proof.openings.to_fri_openings());

        StarkProofChallengesTarget {
            permutation_challenge_sets,
            stark_alphas,
            stark_zeta,
            fri_challenges: challenger.fri_challenges::<C>(
                builder,
                &opening_proof.commit_phase_merkle_caps,
                &opening_proof.final_poly,
                opening_proof.pow_witness,
                &config.fri_config,
            ),
        }
```

Reply:
Good idea! I've done the same for the native version. Having the separate `get_challenges` function is useful for compression, but since we won't need it I agree inlining is nice.